File: PE-common/src/main/java/nl/adaptivity/io/WritableReader.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.io;

import java.io.CharArrayReader;
import java.io.CharArrayWriter;
import java.io.IOException;
import java.io.Reader;
import java.nio.CharBuffer;


/**
 * Created by pdvrieze on 19/11/15.
 */
public class WritableReader extends Reader {

  private Writable mContent;

  private Reader mDelegate;

  public WritableReader(final Writable content) {
    mContent = content;
  }

  private Reader getDelegate() throws IOException {
    if (mDelegate==null) {
      CharArrayWriter caw = new CharArrayWriter();
      mContent.writeTo(caw);
      mDelegate = new CharArrayReader(caw.toCharArray());
      mContent = null; // No longer needed, discard.
    }
    return mDelegate;
  }

  @Override
  public int read(final CharBuffer target) throws IOException {
    return getDelegate().read(target);
  }

  @Override
  public int read() throws IOException {
    return getDelegate().read();
  }

  @Override
  public int read(final char[] cbuf) throws IOException {
    return getDelegate().read(cbuf);
  }

  @Override
  public int read(final char[] cbuf, final int off, final int len) throws IOException {
    return getDelegate().read(cbuf, off, len);
  }

  @Override
  public long skip(final long n) throws IOException {
    return getDelegate().skip(n);
  }

  @Override
  public boolean ready() throws IOException {
    return getDelegate().ready();
  }

  @Override
  public boolean markSupported() {
    try {
      return getDelegate().markSupported();
    } catch (IOException e) {
      return false;
    }
  }

  @Override
  public void mark(final int readAheadLimit) throws IOException {
    getDelegate().mark(readAheadLimit);
  }

  @Override
  public void reset() throws IOException {
    getDelegate().reset();
  }

  @Override
  public void close() throws IOException {
    getDelegate().close();
  }
}

File: PE-common/src/jvmMain/kotlin/nl/adaptivity/io/WritableReader.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.io

import java.io.CharArrayReader
import java.io.CharArrayWriter
import java.io.IOException
import java.io.Reader
import java.nio.CharBuffer


/**
 * Created by pdvrieze on 19/11/15.
 */
class WritableReader(content: Writable) : Reader() {

    private var content: Writable? = content

    private// No longer needed, discard.
    val delegate: Reader by lazy {
        val caw = CharArrayWriter()
        content.writeTo(caw)
        this.content = null
        CharArrayReader(caw.toCharArray())
    }

    @Throws(IOException::class)
    override fun read(target: CharBuffer): Int {
        return delegate.read(target)
    }

    @Throws(IOException::class)
    override fun read(): Int {
        return delegate.read()
    }

    @Throws(IOException::class)
    override fun read(cbuf: CharArray): Int {
        return delegate.read(cbuf)
    }

    @Throws(IOException::class)
    override fun read(cbuf: CharArray, off: Int, len: Int): Int {
        return delegate.read(cbuf, off, len)
    }

    @Throws(IOException::class)
    override fun skip(n: Long): Long {
        return delegate.skip(n)
    }

    @Throws(IOException::class)
    override fun ready(): Boolean {
        return delegate.ready()
    }

    override fun markSupported(): Boolean {
        try {
            return delegate.markSupported()
        } catch (e: IOException) {
            return false
        }

    }

    @Throws(IOException::class)
    override fun mark(readAheadLimit: Int) {
        delegate.mark(readAheadLimit)
    }

    @Throws(IOException::class)
    override fun reset() {
        delegate.reset()
    }

    @Throws(IOException::class)
    override fun close() {
        delegate.close()
    }
}
